### PR TITLE
added more updaters and learning rate for the minimizer framework

### PR DESCRIPTION
**Pull request by yorkerlin**
The idea of Adaptive momentum is
* to increase the `next` gradient if the sign of velocity and the sign of `current` gradient are agreed
* to discount the `next` gradient if the sign of velocity and the sign of `current` gradient are not agreed

### CODE REVIEW

File: src/shogun/optimization/DescendCorrection.h [hunk 1]
```diff
@@ -68,11 +68,12 @@ class DescendCorrection
 	 *
 	 * @param negative_descend_direction the negative descend direction
 	 * @param idx the index of the direction
+	 * @param delta update the change to correct descend direction
 	 * 
 	 * @return corrected descend direction
 	 */
 	virtual float64_t get_corrected_descend_direction(float64_t negative_descend_direction,
-		index_t idx)=0;
+		index_t idx, float64_t& delta)=0;
```
Comment (lisitsyn) [inline src/shogun/optimization/DescendCorrection.h]: Is it necessary for delta to be changed through reference?

File: src/shogun/optimization/NesterovMomentumCorrection.h [hunk 2]
```diff
@@ -92,15 +92,18 @@ class NesterovMomentumCorrection: public MomentumCorrection
...
 	 */
-	virtual float64_t get_corrected_descend_direction(float64_t gradient, index_t idx)
+	virtual float64_t get_corrected_descend_direction(float64_t gradient, index_t idx,
+		float64_t& delta)
 	{
 		REQUIRE(idx>=0 && idx<m_previous_descend_direction.vlen,"The index (%d) is invalid\n", idx);
 		float64_t tmp=m_weight*m_previous_descend_direction[idx];
 		m_previous_descend_direction[idx]=tmp-gradient;
-		return (1.0+m_weight)*m_previous_descend_direction[idx]-tmp;
+		delta=m_previous_descend_direction[idx];
```
Comment (yorkerlin) [inline src/shogun/optimization/NesterovMomentumCorrection.h]: @lisitsyn 
Yes.
For `NesterovMomentumCorrection`, the `return` is different from `delta`.

File: src/shogun/optimization/DescendCorrection.h (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/optimization/DescendCorrection.h]: Yes.
If we allow to return a `vector`, we do not need to use reference.
However, the `delta` is not a `corrected_descend_direction`.

Comment (yorkerlin) [inline src/shogun/optimization/DescendCorrection.h]: The idea of Adaptive momentum is 
* increase the gradient if the sign of `velocity` and the sign of `gradient` are agreed
* discount the gradient  if the sign of `velocity` and the sign of `gradient` are not agreed

**Comment by yorkerlin** [comment]
@lisitsyn 
Locally, momentum methods (including the adaptive momentum method) can be introduced to `RmsProp`, `Adam`, `Adadelta` and `adagrad`.

I have tested the result for `RmsProp` and `Adadelta`.

File: src/shogun/optimization/DescendCorrection.h (hunk 1 above)
Comment (lisitsyn) [inline src/shogun/optimization/DescendCorrection.h]: Can we return pair of values?

**Comment by yorkerlin** [comment]
@lisitsyn 
@vigsterkr 
It seems `CMap` (including `CStringMap`) is `not` serializable.

**Comment by yorkerlin** [comment]
@lisitsyn 
Since `CMap` and `CStringMap` do not support serialization, I disable  serialization for `MinimizerContext`.

**Comment by yorkerlin** [comment]
@lambday @lisitsyn 
Currently, most of classes in the minimizer framework are not subclasses of SGObject.
If in future we allow all these classes in the minimizer framework become  subclasses of SGObject,
the `serialization` issue will be solved.
If not, I have to use `CStringMap` (a subclass of `CMap`) for serialization. 
In that case, do you guys plan to support serialization for `CMap`?

**Comment by yorkerlin** [comment]
Errors are not related to this PR.
I merge  this.

**Pull request merged**